### PR TITLE
CloudXNS: client isolation.

### LINKED ISSUE #619: Wrong host record in CloudXNS
When I use [Caddy](https://caddyserver.com/) as an HTTP server in DNS challenge mode, it cannot get certificate of my domain hosted by [CloudXNS](https://www.cloudxns.net/).

It seems that there is something wrong with record name generated by Lego.
The host record is set to `_acme-challenge.example.com`.
qq20180816-175531 2x

But CloudXNS will resolve it to `_acme-challenge.example.com.example.com`.
qq20180816-174143 2x

According to [the document of CloudXNS](https://www.cloudxns.net/Public/Doc/CloudXNS_api2.0_doc_zh-cn.zip) (Chinese version only, sad...), the key `host` **SHOULD NOT** contain the domain again.
For example `_acme-challenge` means `_acme-challenge.example.com` and `_acme-challenge.foo` means `_acme-challenge.foo.example.com`.

Since I don't know GO, I cannot bring a PR. But I think the problem is here:
[https://github.com/xenolf/lego/blob/master/providers/dns/cloudxns/cloudxns.go#L139](https://github.com/xenolf/lego/blob/master/providers/dns/cloudxns/cloudxns.go#L139)

Hoping you can fix it. Thanks a lot.

### PR DESCRIPTION
**Pull request by ldez**
Client isolation to facilitate testing.

Fixes #619

### CODE REVIEW

File: providers/dns/cloudxns/cloudxns.go [hunk 1]
```diff
@@ -1,32 +1,45 @@
...
+	APIKey     string
+	SecretKey  string
+	TTL        int
+	HTTPClient *http.Client
+}
+
+// NewDefaultConfig returns a default configuration for the DNSProvider
+func NewDefaultConfig() *Config {
+	client := acme.HTTPClient
+	client.Timeout = time.Second * time.Duration(env.GetOrDefaultInt("CLOUDXNS_HTTP_TIMEOUT", 30))
```
Comment (wyattjoh) [inline providers/dns/cloudxns/cloudxns.go]: Is this not modifying the default/package level client?

File: providers/dns/cloudxns/client.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+		HttClient: &http.Client{},
+		BaseURL:   defaultBaseURL,
+	}, nil
+}
+
+// Client CloudXNS client
+type Client struct {
+	apiKey    string
+	secretKey string
+	HttClient *http.Client
```
Comment (wyattjoh) [inline providers/dns/cloudxns/client.go]: Is there a reason this is `HttClient` vs `HTTPClient`?

File: providers/dns/cloudxns/client.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+	req.Header.Set("API-KEY", c.apiKey)
+	req.Header.Set("API-REQUEST-DATE", requestDate)
+	req.Header.Set("API-HMAC", c.hmac(url, requestDate, string(body)))
+	req.Header.Set("API-FORMAT", "json")
+
+	return req, nil
+}
+
+func (c *Client) hmac(url, date, body string) string {
+	sum := md5.Sum([]byte(c.apiKey + url + body + date + c.secretKey))
```
Comment (wyattjoh) [inline providers/dns/cloudxns/client.go]: Could use a comment to explain what's going on here, a bit cryptic.

File: providers/dns/cloudxns/client_test.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func handlerMock(method string, response *apiResponse, data interface{}) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		if req.Method != method {
+			content, err := json.Marshal(apiResponse{
+				Code:    666,
```
Comment (wyattjoh) [inline providers/dns/cloudxns/client_test.go]: 666? Needs comments as to why this is the code.

File: providers/dns/cloudxns/client.go (hunk 3 above)
Comment (ldez) [reply providers/dns/cloudxns/client.go]: It's not my code: I just move the code from `DNSProvider`

File: providers/dns/cloudxns/client.go (hunk 2 above)
Comment (ldez) [reply providers/dns/cloudxns/client.go]: good catch

File: providers/dns/cloudxns/cloudxns.go (hunk 1 above)
Comment (ldez) [reply providers/dns/cloudxns/cloudxns.go]: no because I use a copy of `acme.HTTPClient`

File: providers/dns/cloudxns/client_test.go (hunk 4 above)
Comment (ldez) [reply providers/dns/cloudxns/client_test.go]: It's just a dumb code

**Pull request merged**